data/blog/tailscale-sucks.mdx

```diff
@@ -0,0 +1,456 @@
+</figure>
+
+Finally, I powered on the robot, and... Success! The serial console lights up
+with kernel logs as the SoC boots up. With a little more fiddling which I'll
+gloss over (see [Valetudo's docs][valetudo-install]), I have a root shell on the
+robot's OS, which as it turns out is an OpenWRT fork.
+
+From here, it's a simple matter of backing up irreplaceable files (private keys,
+device IDs, sensor calibration data from the factory), and installing a patched
+firmware that breaks the cloud connection. A little NAND flashing and rebooting
+later, and my robot is now a standalone system, with an SSH server, and Valetudo
+providing a replacement for the vendor app's functionality!
+
+<figure className="float-none text-center sm:pl-4">
+  <Image
+    src="/images/tailscale-sucks/valetudo.jpg"
+    width="626"
+    height="392"
+    alt="Screenshot of the Valetudo web app. On the left is a lidar map of a house, segmented by color into rooms. On the right are various settings like vacuum intensity and operating mode."
+  />
+  <figcaption>
+    Valetudo is so much more responsive than the manufacturer's app, it's not even funny. Turns out,
+    computers are fast when you don't send every request via the cloud!
+  </figcaption>
+</figure>
+
+At this point, my robot can do everything that it could do on stock firmware,
+with no cloud tethering. Well... Almost everything. Valetudo is serving its
+control API and web UI over HTTP on my LAN, which is fine, but I'd prefer to
+have some form of secure access. For all its downsides, the cloud connection did
+provide that, whereas now we're stuck with HTTP over local network. Or are we?
+
+## Getting Tailscale on the robot
+
+Having done all this legwork, the robot is now a stripped down linux computer
+that I control. That means I should be able to get Tailscale running on it, and
+give myself a secure way to access Valetudo, and for Valetudo to dial out to my
+home automation stack.
+
+SSHing into the robot, we find a barebones linux OS, consisting mostly of
+busybox and a few addon binaries. Most of the filesystem is a read-only
+SquashFS, but we can write to `/tmp` which is discarded on reboot, or `/data`
+which is persisted across reboots and firmware updates. One wrinkle is that the
+firmware doesn't ship `iptables` (and the kernel doesn't have nftables compiled
+in), so we'll have to run tailscale in [userspace networking mode][ts-user].
+That's probably for the best anyway, as I feel more comfortable leaving the OS
+network stack and DNS config alone on such an esoteric platform.
+
+The robot has a generous 240MiB free space in /data, so we can be a bit lazy and
+just grab static arm64 Tailscale binaries from [the package server][pkgs] and
+use those. They are not optimized for size, so we'll end up eating about 40MiB
+of the free space. If you want to be more economical, you can build a stripped
+down multicall binary from [Tailscale's source code repo][ts-source], by running:
```

Comment:
link https://tailscale.com/kb/1207/small-tailscale/ in here somewhere?

Reply:
Done.